### PR TITLE
[Merged by Bors] - chore: golf using `filter_upwards`

### PR DESCRIPTION
**Pull request by grunweg**
This is presumably not exhaustive, but covers about a hundred instances.

Style opinions (e.g., why a particular change is great/not a good idea) are very welcome; I'm still forming my own.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Probability/Kernel/IntegralCompProd.lean [hunk 1]
```diff
@@ -99,7 +99,7 @@ theorem hasFiniteIntegral_compProd_iff ⦃f : β × γ → E⦄ (h1f : StronglyM
     rw [← and_congr_right_iff, and_iff_right_of_imp h1]
   rw [this]
   · intro h2f; rw [lintegral_congr_ae]
-    refine' h2f.mp _; apply eventually_of_forall; intro x hx; dsimp only
+    refine' h2f.mp _; filter_upwards with x hx
```
Comment (loefflerd) [inline Mathlib/Probability/Kernel/IntegralCompProd.lean:102]:
```suggestion
    filter_upwards [h2f] with x hx
```
IMHO, this is the true power of `filter_upwards`; when you're using it without a [...] argument the gain over `eventually_of_forall` is minor. (The rest of this proof can also golfed considerably but that's not the point of this PR.)

File: Mathlib/MeasureTheory/Constructions/Prod/Integral.lean [hunk 2]
```diff
@@ -473,9 +473,9 @@ theorem integral_prod (f : α × β → E) (hf : Integrable f (μ.prod ν)) :
   · exact isClosed_eq continuous_integral continuous_integral_integral
   · rintro f g hfg - hf; convert hf using 1
     · exact integral_congr_ae hfg.symm
-    · refine' integral_congr_ae _
-      refine' (ae_ae_of_ae_prod hfg).mp _
-      apply eventually_of_forall; intro x hfgx
+    · refine integral_congr_ae ?_
+      refine (ae_ae_of_ae_prod hfg).mp ?_
+      filter_upwards with x hfgx
       exact integral_congr_ae (ae_eq_symm hfgx)
```
Comment (loefflerd) [inline Mathlib/MeasureTheory/Constructions/Prod/Integral.lean:479]:
```suggestion
    · apply integral_congr_ae
      filter_upwards [ae_ae_of_ae_prod hfg] with x hfgx using integral_congr_ae (ae_eq_symm hfgx)```

File: Mathlib/Analysis/Calculus/Rademacher.lean [hunk 3]
```diff
@@ -124,15 +124,15 @@ theorem integral_inv_smul_sub_mul_tendsto_integral_lineDeriv_mul'
   have K_compact : IsCompact K := IsCompact.cthickening h'f
   apply tendsto_integral_filter_of_dominated_convergence
       (K.indicator (fun x ↦ (C * ‖v‖) * ‖g x‖))
-  · apply eventually_of_forall (fun t ↦ ?_)
+  · filter_upwards with t
     apply AEStronglyMeasurable.mul ?_ hg.aestronglyMeasurable
     apply aestronglyMeasurable_const.smul
     apply AEStronglyMeasurable.sub _ hf.continuous.measurable.aestronglyMeasurable
     apply AEMeasurable.aestronglyMeasurable
     exact hf.continuous.measurable.comp_aemeasurable' (aemeasurable_id'.add_const _)
```
Comment (loefflerd) [inline Mathlib/Analysis/Calculus/Rademacher.lean:132]: Off-topic: can this be golfed with `measurability` or `fun_prop`?

**Review by loefflerd — APPROVED** [review]
This PR is definitely an improvement but I saw a couple of points where it can be improved further using the `filter_upwards [foo] with bar` syntax. There might be more, so have a look for cases where `Eventually.mp` (or `Eventually.and` etc) is used just before a `filter_upwards`.

**Comment by grunweg** [comment]
Thanks for the hint: I pushed a few more golfs, including looking for occurrences of Eventually.{mp,and}.
awaiting-review

File: Mathlib/Analysis/Calculus/Rademacher.lean (hunk 3 above)
Comment (grunweg) [reply Mathlib/Analysis/Calculus/Rademacher.lean:132]: This seems tempting, I agree. Off the shelf, neither of them works.

File: Mathlib/Analysis/Convolution.lean [hunk 4]
```diff
@@ -1073,8 +1073,8 @@ theorem HasCompactSupport.hasFDerivAt_convolution_right (hcg : HasCompactSupport
   -- `refine' hasFDerivAt_integral_of_dominated_of_fderiv_le zero_lt_one h1 _ (h2 x₀) _ _ _`
   -- but it failed; surprisingly, `apply` works
   apply hasFDerivAt_integral_of_dominated_of_fderiv_le zero_lt_one h1 _ (h2 x₀)
-  · refine' eventually_of_forall fun t x hx =>
-      (hcg.fderiv 𝕜).convolution_integrand_bound_right L' (hg.continuous_fderiv le_rfl)
+  · filter_upwards with t x hx
+    exact (hcg.fderiv 𝕜).convolution_integrand_bound_right L' (hg.continuous_fderiv le_rfl)
```
Comment (loefflerd) [inline Mathlib/Analysis/Convolution.lean:1077]: `using`?

File: Mathlib/Analysis/Fourier/AddCircle.lean [hunk 5]
```diff
@@ -411,7 +411,8 @@ theorem fourierCoeff_liftIoc_eq {a : ℝ} (f : ℝ → ℂ) (n : ℤ) :
     fourierCoeffOn (lt_add_of_pos_right a hT.out) f n := by
   rw [fourierCoeffOn_eq_integral, fourierCoeff_eq_intervalIntegral, add_sub_cancel' a T]
   congr 1
-  refine' intervalIntegral.integral_congr_ae (ae_of_all _ fun x hx => _)
+  apply intervalIntegral.integral_congr_ae
+  filter_upwards with x hx
```
Comment (loefflerd) [inline Mathlib/Analysis/Fourier/AddCircle.lean:415]: I'm not sure this is an improvement?

File: Mathlib/Analysis/Fourier/FourierTransform.lean [hunk 6]
```diff
@@ -175,11 +175,12 @@ theorem fourierIntegral_continuous [FirstCountableTopology W] (he : Continuous e
...
   apply continuous_of_dominated
   · exact fun w => ((fourier_integral_convergent_iff he hL w).mp hf).1
-  · refine' fun w => ae_of_all _ fun v => _
+  · intro w; filter_upwards with v
     · rw [norm_smul, Complex.norm_eq_abs, abs_coe_circle, one_mul]
   · exact hf.norm
   · rw [continuous_induced_rng] at he
-    refine' ae_of_all _ fun v => (he.comp (continuous_ofAdd.comp _)).smul continuous_const
+    filter_upwards with v
+    apply (he.comp (continuous_ofAdd.comp _)).smul continuous_const
```
Comment (loefflerd) [inline Mathlib/Analysis/Fourier/FourierTransform.lean:183]: these are both pretty marginal

File: Mathlib/Analysis/Fourier/FourierTransformDeriv.lean [hunk 7]
```diff
@@ -106,11 +106,11 @@ theorem hasFDerivAt_fourier [CompleteSpace E] [MeasurableSpace V] [BorelSpace V]
...
       apply aux0.comp_aestronglyMeasurable aux1
   have h4 : (∀ᵐ v ∂μ, ∀ (w' : W), w' ∈ Metric.ball w 1 → ‖F' w' v‖ ≤ B v) := by
-    refine ae_of_all _ (fun v w' _ ↦ ?_)
+    filter_upwards with v w' _
     exact norm_fderiv_fourier_transform_integrand_right_le L f v w'
   have h5 : Integrable B μ := by simpa only [← mul_assoc] using hf'.const_mul (2 * π * ‖L‖)
-  have h6 : ∀ᵐ v ∂μ, ∀ w', w' ∈ Metric.ball w 1 → HasFDerivAt (fun x ↦ F x v) (F' w' v) w' :=
-    ae_of_all _ (fun v w' _ ↦ hasFDerivAt_fourier_transform_integrand_right L f v w')
+  have h6 : ∀ᵐ v ∂μ, ∀ w', w' ∈ Metric.ball w 1 → HasFDerivAt (fun x ↦ F x v) (F' w' v) w' := by
+    filter_upwards with v w' _ using hasFDerivAt_fourier_transform_integrand_right L f v w'
```
Comment (loefflerd) [inline Mathlib/Analysis/Fourier/FourierTransformDeriv.lean:113]: here the `filter_upwards` version is longer (and presumably also slower) than the original

File: Mathlib/Analysis/MellinTransform.lean [hunk 8]
```diff
@@ -354,7 +354,8 @@ theorem mellin_hasDerivAt_of_isBigO_rpow [CompleteSpace E] [NormedSpace ℂ E] {
       ⟨min w w', lt_min hw1 hw1', (min_le_right _ _).trans_lt hw2', (min_le_left _ _).trans_lt hw2⟩
   let bound : ℝ → ℝ := fun t : ℝ => (t ^ (s.re + v - 1) + t ^ (s.re - v - 1)) * |log t| * ‖f t‖
   have h1 : ∀ᶠ z : ℂ in 𝓝 s, AEStronglyMeasurable (F z) (volume.restrict <| Ioi 0) := by
-    refine' eventually_of_forall fun z => AEStronglyMeasurable.smul _ hfc.aestronglyMeasurable
+    filter_upwards with z
+    refine AEStronglyMeasurable.smul ?_ hfc.aestronglyMeasurable
```
Comment (loefflerd) [inline Mathlib/Analysis/MellinTransform.lean:358]: again, this doesn't achieve much

File: Mathlib/Analysis/SpecialFunctions/Gamma/Basic.lean [hunk 9]
```diff
@@ -105,8 +105,8 @@ theorem GammaIntegral_convergent {s : ℂ} (hs : 0 < s.re) :
     exact ContinuousAt.comp this continuous_ofReal.continuousAt
   · rw [← hasFiniteIntegral_norm_iff]
     refine' HasFiniteIntegral.congr (Real.GammaIntegral_convergent hs).2 _
-    refine' (ae_restrict_iff' measurableSet_Ioi).mpr (ae_of_all _ fun x hx => _)
-    dsimp only
+    apply (ae_restrict_iff' measurableSet_Ioi).mpr
```
Comment (loefflerd) [inline Mathlib/Analysis/SpecialFunctions/Gamma/Basic.lean:108]: `rw [ae_restrict_iff' measurableSet_Ioi]`?

File: Mathlib/Analysis/SpecialFunctions/Gamma/Beta.lean [hunk 10]
```diff
@@ -334,7 +334,8 @@ theorem approx_Gamma_integral_tendsto_Gamma_integral {s : ℂ} (hs : 0 < re s) :
       Measure.restrict_restrict_of_subset Ioc_subset_Ioi_self]
   -- f is uniformly bounded by the Gamma integrand
   · intro n
-    refine' (ae_restrict_iff' measurableSet_Ioi).mpr (ae_of_all _ fun x hx => _)
+    refine' (ae_restrict_iff' measurableSet_Ioi).mpr ?_
```
Comment (loefflerd) [inline Mathlib/Analysis/SpecialFunctions/Gamma/Beta.lean:337]: `rw [ae_restrict_iff' measurableSet_Ioi]`

File: Mathlib/Analysis/SpecialFunctions/Integrals.lean [hunk 11]
```diff
@@ -882,7 +882,8 @@ theorem integral_sqrt_one_sub_sq : ∫ x in (-1 : ℝ)..1, sqrt (1 - x ^ 2 : ℝ
           (integral_comp_mul_deriv (fun x _ => hasDerivAt_sin x) continuousOn_cos
             (by continuity)).symm
     _ = ∫ x in (-(π / 2))..(π / 2), cos x ^ 2 := by
-          refine integral_congr_ae (MeasureTheory.ae_of_all _ fun _ h => ?_)
+          apply integral_congr_ae
+          filter_upwards with _ h
```
Comment (loefflerd) [inline Mathlib/Analysis/SpecialFunctions/Integrals.lean:886]: dubious

File: Mathlib/Analysis/SpecialFunctions/NonIntegrable.lean [hunk 12]
```diff
@@ -92,7 +92,7 @@ theorem not_integrableOn_of_tendsto_norm_atTop_of_deriv_isBigO_filter_aux
     _ ≤ ∫ x in Ι c d, C * ‖g x‖ :=
       set_integral_mono_on hfi.norm.def (hgi.mono_set hsub') measurableSet_uIoc hg
     _ ≤ ∫ x in k, C * ‖g x‖ := by
-      apply set_integral_mono_set hgi
+      exact set_integral_mono_set hgi
```
Comment (loefflerd) [inline Mathlib/Analysis/SpecialFunctions/NonIntegrable.lean:95]: this has nothing to do with the PR (and there are situations where `apply foo` is faster than `exact foo` even when the latter does work)

File: Mathlib/MeasureTheory/Decomposition/RadonNikodym.lean [hunk 13]
```diff
@@ -80,7 +80,7 @@ lemma rnDeriv_pos [HaveLebesgueDecomposition μ ν] (hμν : μ ≪ ν) :
     ∀ᵐ x ∂μ, 0 < μ.rnDeriv ν x := by
   rw [← Measure.withDensity_rnDeriv_eq _ _  hμν,
     ae_withDensity_iff (Measure.measurable_rnDeriv _ _), Measure.withDensity_rnDeriv_eq _ _  hμν]
-  exact ae_of_all _ (fun x hx ↦ lt_of_le_of_ne (zero_le _) hx.symm)
+  filter_upwards with x hx using lt_of_le_of_ne (zero_le _) hx.symm
```
Comment (loefflerd) [inline Mathlib/MeasureTheory/Decomposition/RadonNikodym.lean:83]: dubious

File: Mathlib/MeasureTheory/Decomposition/RadonNikodym.lean [hunk 14]
```diff
@@ -193,13 +193,13 @@ theorem rnDeriv_restrict (μ ν : Measure α) [SigmaFinite μ] [SigmaFinite ν]
     {s : Set α} (hs : MeasurableSet s) :
     (μ.restrict s).rnDeriv ν =ᵐ[ν] s.indicator (μ.rnDeriv ν) := by
   rw [← withDensity_indicator_one hs]
-  refine (rnDeriv_withDensity_left ?_ ?_ ?_).trans (ae_of_all _ (fun x ↦ ?_))
+  refine (rnDeriv_withDensity_left ?_ ?_ ?_).trans ?_
```
Comment (loefflerd) [inline Mathlib/MeasureTheory/Decomposition/RadonNikodym.lean:196]: this change (+ its twin at line 202) seem dubious (but the one in between is good)

File: Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean [hunk 15]
```diff
@@ -1697,9 +1697,8 @@ theorem _root_.aestronglyMeasurable_of_tendsto_ae {ι : Type*} [PseudoMetrizable
     refine ⟨closure (⋃ i, t i), .closure <| .iUnion t_sep, ?_⟩
     filter_upwards [ae_all_iff.2 ht, lim] with x hx h'x
     apply mem_closure_of_tendsto (h'x.comp hv)
-    refine eventually_of_forall fun n => ?_
-    apply mem_iUnion_of_mem n
-    exact hx n
+    filter_upwards with n
+    exact mem_iUnion_of_mem n (hx n)
```
Comment (loefflerd) [inline Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean:1701]: `using`?

File: Mathlib/MeasureTheory/Integral/DominatedConvergence.lean [hunk 16]
```diff
@@ -235,10 +235,10 @@ theorem hasSum_intervalIntegral_of_summable_norm [Countable ι] {f : ι → C(
   apply hasSum_integral_of_dominated_convergence
     (fun i (x : ℝ) => ‖(f i).restrict ↑(⟨uIcc a b, isCompact_uIcc⟩ : Compacts ℝ)‖)
     (fun i => (map_continuous <| f i).aestronglyMeasurable)
-  · refine fun i => ae_of_all _ fun x hx => ?_
+  · intro i; filter_upwards with x hx
     apply ContinuousMap.norm_coe_le_norm ((f i).restrict _) ⟨x, _⟩
     exact ⟨hx.1.le, hx.2⟩
-  · exact ae_of_all _ fun x _ => hf_sum
+  · filter_upwards with x _ using hf_sum
```
Comment (loefflerd) [inline Mathlib/MeasureTheory/Integral/DominatedConvergence.lean:241]: no real gain here

File: Mathlib/MeasureTheory/Integral/IntervalIntegral.lean [hunk 17]
```diff
@@ -1098,7 +1098,8 @@ theorem intervalIntegral_pos_of_pos_on {f : ℝ → ℝ} {a b : ℝ} (hfi : Inte
   have h₀ : 0 ≤ᵐ[volume.restrict (uIoc a b)] f := by
     rw [EventuallyLE, uIoc_of_le hab.le]
     refine' ae_restrict_of_ae_eq_of_ae_restrict Ioo_ae_eq_Ioc _
-    exact (ae_restrict_iff' measurableSet_Ioo).mpr (ae_of_all _ fun x hx => (hpos x hx).le)
+    apply (ae_restrict_iff' measurableSet_Ioo).mpr
+    filter_upwards with x hx using (hpos x hx).le
```
Comment (loefflerd) [inline Mathlib/MeasureTheory/Integral/IntervalIntegral.lean:1102]: `apply foo.mpr` is usually better expressed as `rw [foo]`

File: Mathlib/NumberTheory/ZetaFunction.lean [hunk 18]
```diff
@@ -538,7 +538,8 @@ theorem mellin_zetaKernel₁_eq_tsum {s : ℂ} (hs : 1 / 2 < s.re) :
       ((t : ℂ) ^ (s - 1) * zetaKernel₁ t) :=
     (ae_restrict_iff' hm).mpr (ae_of_all _ fun t ht => h_sum0 ht)
   have h_sum : ∀ᵐ t : ℝ ∂volume.restrict (Ioi 0), Summable fun n : ℕ => bd n t := by
-    refine (ae_restrict_iff' hm).mpr (ae_of_all _ fun t ht => ?_)
+    apply (ae_restrict_iff' hm).mpr
+    filter_upwards with t ht
```
Comment (loefflerd) [inline Mathlib/NumberTheory/ZetaFunction.lean:542]: no gain

File: Mathlib/Probability/Kernel/IntegralCompProd.lean [hunk 19]
```diff
@@ -113,12 +113,11 @@ theorem hasFiniteIntegral_compProd_iff' ⦃f : β × γ → E⦄
...
   · apply eventually_congr
-    filter_upwards [ae_ae_of_ae_compProd h1f.ae_eq_mk.symm]
-    intro x hx
+    filter_upwards [ae_ae_of_ae_compProd h1f.ae_eq_mk.symm] with x hx
     exact hasFiniteIntegral_congr hx
   · apply hasFiniteIntegral_congr
-    filter_upwards [ae_ae_of_ae_compProd h1f.ae_eq_mk.symm] with _ hx using
-      integral_congr_ae (EventuallyEq.fun_comp hx _)
+    filter_upwards [ae_ae_of_ae_compProd h1f.ae_eq_mk.symm] with _ hx
+    exact integral_congr_ae (EventuallyEq.fun_comp hx _)
```
Comment (loefflerd) [inline Mathlib/Probability/Kernel/IntegralCompProd.lean:120]: seems like a strange edit, was this really intended?

**Review by loefflerd — COMMENTED** [review]
I've checked through all of these now. Most of them look great. I have, however, flagged up a few instances where I think you've been a bit trigger-happy. In particular, if `eventually_of_forall` or `ae_of_all` is one ingredient in a long compound proof term (as argument to `refine` or `exact`), using the `filter_upwards` tactic requires breaking this into two tactic calls, and in many cases the result is longer than the original (as well as being, I suspect, fractionally slower to compile).

File: Mathlib/Analysis/SpecialFunctions/Gamma/Beta.lean (hunk 10 above)
Comment (grunweg) [reply Mathlib/Analysis/SpecialFunctions/Gamma/Beta.lean]: This worked and is even nicer.

File: Mathlib/Analysis/SpecialFunctions/Gamma/Basic.lean (hunk 9 above)
Comment (grunweg) [reply Mathlib/Analysis/SpecialFunctions/Gamma/Basic.lean:108]: This doesn't work, so I've left it as-is.

**Comment by grunweg** [comment]
Thank you, all fixed. (Sorry for going overboard - I rather wanted you to tell me "this is not better" than the other way. Now I know.)
awaiting-review

**Review by loefflerd — APPROVED** [review]
This looks good now. There are surely further golfs that are possible; but this PR certainly is a substantial improvement on the existing code, and given that it touches rather a lot of files, it will rot quickly if we try to make it perfect.

maintainer merge

**Comment by fpvandoorn** [comment]
This looks good, thanks!

bors merge

**Pull request closed without merging**